.github/workflows/test.yml

```diff
@@ -11,152 +11,127 @@ on:
+          }
+          if ($env:RUST_TOOLCHAIN -like "*i686*") {
+            rustup target add i686-pc-windows-gnu ;
+            echo "CARGO_BUILD_FLAGS=--target=i686-pc-windows-gnu" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append ;
+            echo "LIBCLANG_PATH=C:\msys64\mingw32\bin" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append ;
+            echo "C:\msys64\mingw32\bin" | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append ;
+            echo "$(Rscript.exe -e 'cat(normalizePath(R.home()))')\bin\i386"  | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append ;
+
+          }
+        env: 
+          RUST_TOOLCHAIN: ${{ matrix.config.rust-version }}
+          
+
+      # MacOS configurations, mainly `llvm` and path to `libclang`
+      - name: Configure MacOs
+        if: runner.os == 'macOS'
+        run: | 
+          brew install llvm r
+          echo "LIBCLANG_PATH=$(brew --prefix llvm)/lib" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
+          
+
+      # This is required for ubuntu r-devel
+      # 'Del alias:R' removes 'R' alias which prevents running R 
+      - name: Configure Linux
+        if: runner.os == 'linux'
+        run: |
+          Del alias:R
+          echo "LD_LIBRARY_PATH=$(R -s -e 'cat(normalizePath(R.home()))')/lib" | Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append
+
+      # Build and emit bindings to `./generated_bindings`
+      - name: Build & Emit bindings
         id: build
         run: |
-          cargo build -vv --features use-bindgen 
+          cargo build -vv --features use-bindgen $env:CARGO_BUILD_FLAGS
         env:
-          LIBCLANG_PATH: C:/msys64/mingw64/bin
           LIBRSYS_BINDINGS_OUTPUT_PATH: generated_bindings
-      - name: Build i686
-        id: build_i686
-        run: |
-          cargo build -vv --features use-bindgen --target i686-pc-windows-gnu
-        env:
-          LIBCLANG_PATH: C:/msys64/mingw64/bin
-          LIBRSYS_BINDINGS_OUTPUT_PATH: generated_bindings_i686
-          LIBRSYS_LIBCLANG_INCLUDE_PATH: C:/msys64/mingw32/i686-w64-mingw32/include/
+
+      # Test the result of previous step
       - name: Run tests
-        run: cargo test --features use-bindgen -- --nocapture --test-threads=1
+        id: test
+        run: |
+          cargo test -vv --features use-bindgen $env:CARGO_BUILD_FLAGS -- --nocapture --test-threads=1
+
+
+      # If success, upload bindings
       - name: Upload generated bindings
-        if: steps.build.outcome == 'success'
+        if: steps.build.outcome == 'success' && steps.test.outcome == 'success' && startsWith(matrix.config.rust-version, 'stable')
         uses: actions/upload-artifact@main
         with:
-          name: Windows generated bindings
+          name: ${{ matrix.config.os }} ${{ matrix.config.rust-version }} generated bindings
```

Comment:
The generated bindings need to have the R version in the name also.

Reply:
Artifact names are now identical to job names